drivers: mmc: add SD support for Command Queueing

P33M opened the pull request:
Application class A2 cards require CQ to be enabled to realise their stated performance figures. Add support to enable/disable card CQ via the Performance Enhancement extension register, and cater for the slight differences in command set versus eMMC.

--

Tested using a 64GB Sandisk Extreme U3 class A2 card. Typical performance uplift is ~2.5x for random 4k read, and 4x for random 4k write. There are sporadic instances of errors (usually command timeouts) that result in the queue being flushed, but I've not been able to identify a pattern/set of commands that cause it.

Please don't merge immediately. I want a testing thread on the forums to flush out anything that's going to be fatal to your root filesystem first...

P33M committed:
drivers: mmc: add SD support for Command Queueing

Application class A2 cards require CQ to be enabled to realise their
stated performance figures. Add support to enable/disable card CQ via
the Performance Enhancement extension register, and cater for the slight
differences in command set versus eMMC.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

popcornmix commented:
You didn't mention it, but I assume this is Pi5 only?
Will the command queueing be disabled gracefully on non-supporting pi models?

P33M commented:
Yes, for us this is Pi 5 only. There should be no difference in behaviour for a) non-A2 cards and/or b) host controllers that choose to use HSQ.

P33M commented:
One thing I have noticed it the activity LED on Pi 5 no longer blinks in a sensible fashion - only when IO starts/stops overall.

P33M committed:
drivers: sdhci-brcmstb: set CQE timer clock frequency

CQHCI keeps track of tags in flight with internal timers, so the clock
frequency driving the timer needs to be specified. The config registers
default to 0 (100kHz) which means timeouts will be significantly shorter
than they should be. Assume the timer clock comes from the controller
base clock.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
DTS: bcm2712: emmc2 clock frequency is 200MHz

The bootloader sets this as the controller base clock.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
drivers: mmc: preallocate a block for SD extension register accesses

The Performance Extension register is regularly accessed in a hot path
to do write cache flushes. Don't invoke kmalloc/kfree for every access,
preallocate a 512B buffer for this purpose.

Also remove an unused alloc in sd_enable_cache().

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
drivers: mmc: trigger activity LED when CQE is active

Add a LED_FULL trigger equivalent to mmc_start_request() in
mmc_cqe_start_req(), otherwise it stays off forever.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M committed:
DTS: bcm2712: defer SDIO1 CQE selection to the board dtb

Command Queueing on Linux with SD cards is new, with the potential to cause
filesystem corruption due to bugs in software or card firmware(s).

Default to disabled unless a) we know the interface connects to an eMMC
device or b) we want it switched on.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>

P33M commented:
Fixed the random timeouts - they were actually queue flushes that ended up doing an eMMC ext_csd write instead of an SD expansion register write that resulted in an "illegal command" response.

As CQ now disabled by default on Pi 5 it's safe to go in rpi-update.

pelwell commented on arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts:
```diff
@@ -819,6 +819,7 @@ spi10_cs_pins: &spi10_cs_gpio1 {};
 		random = <&random>, "status";
 		rtc = <&rpi_rtc>, "status";
 		rtc_bbat_vchg = <&rpi_rtc>, "trickle-charge-microvolt:0";
+		sd_cqe = <&sdio1>, "supports-cqe";
```
If the aim is to create a boolean (empty property when true, absent when empty) then `"supports-cqe?"` is the correct syntax.

P33M replied:
Yes, the property only has to exist for it to have an effect

P33M commented:
Yes. If the card supports the feature and it's enabled you get the line `mmc0: Command Queue Engine enabled` in dmesg.
